### PR TITLE
CBOR validation fixes

### PR DESCRIPTION
**Pull request by ericseppanen**
A collection of CBOR validation fixes.

The map fix has an open question (see the comment with "REVIEW NEEDED"); I don't really understand the control flow so I'm not sure whether the fix needs to be implemented in more than one place.

Homogeneous arrays (e.g. `[ * int]`) still don't validate (though they didn't before my record-array change, so I don't think I made anything worse than it was before).  I'll keep looking at it.

### CODE REVIEW

**Comment by ericseppanen** [comment]
I'd also like your opinion on what to do about JSON; some of the unit tests would be useful and at least bug #25 affects JSON validation as well.

While I'm neck-deep in the CBOR code I'd prefer to make as much progress on that code as possible, but it will be too easy to forget to go back and make the corresponding fixes.

Should I hold off on sending PRs until I can achieve CBOR/JSON parity?  Should I just send the CBOR fixes as fast as I can write them, and then try to circle back and do JSON fixes later?

**Comment by anweiss** [comment]
Hey @ericseppanen, definitely keep pushing up changes to the CBOR code as it's a bit out of date. Much appreciated! Also, take a look at #31 to get an idea as to how I was thinking of approaching the array validation issue. The code is by no means clean, but it works. In the case of `[ *a: int, b: tstr ]`, because of the ambiguous `*`, I'm just disregarding all group entries with an occurrence indicator altogether. I've updated the README to encourage users not to define arrays like that for validating JSON.

To clean up the validation code, I also like what you were starting to do in [this commit](https://github.com/ericseppanen/cddl/commit/abd5617c19f88d53d2a2a9d65bec25306dfb5dd5); this notion of a `TetheredType`. I could definitely use some help cleaning up the validation logic because as you've seen, it's super confusing.

File: src/validation/cbor.rs [hunk 1]
```diff
@@ -675,14 +696,32 @@ impl Validator<Value> for CDDL {
...
-      ),
+      GroupEntry::TypeGroupname { ge: tge, span } => {
+        if is_type_prelude(&tge.name.ident) {
+          // Substitute a new AST node for the groupentry validation.
+          // FIXME: this seems like an awkward thing to do.
+          self.validate_type2(
+            &Type2::Typename {
+              ident: tge.name.clone(),
+              generic_arg: tge.generic_arg.clone(),
+              span: span.clone(),
```
Comment (anweiss) [inline src/validation/cbor.rs]:
```suggestion
              span,
```

**Comment by anweiss** [comment]
Hey @ericseppanen, I think I've been able to implement homogenous array validation in #31. Feel free to give that same  approach a whirl as part of this PR.

**Comment by ericseppanen** [comment]
I did have a look at #31 and borrowed what I think is the core of the fix in the commit titled `cbor: make validation of homogenous arrays work`.  I'm not confident that I fully understand parts of it, though.

I think I'm done attaching new validation fixes to this PR. I've fixed all of the basic CBOR failures that were blockers for me right now (though more remain; see the broken tests in `cbor_tests.rs`.)

At this point, the "cognitive complexity" issue is more worrisome than the individual bugs.  As someone approaching this code for the first time, I'm finding the code really hard to understand, and I'm never really sure that my fixes are correct.

I know the README says this project is "a personal learning exercise," but it's a really useful crate to me, so I'd love to see it become more complete and trustworthy.

Are you open to a broader discussion on refactoring the validator(s)?

File: src/validation/cbor.rs (hunk 1 above)
Comment (ericseppanen) [reply src/validation/cbor.rs]: This makes the compiler sad:
```
expected tuple, found reference
help: consider dereferencing the borrow: `span: *span`
```
Span is Copy, so I think that `*span` and `span.clone()` are effectively the same thing.  (And I'm happy to follow whatever style you prefer)

If in the future Span might become a nontrivial non-Copy type, I think `*span` is likely to trigger "move out of borrowed context" errors.

Comment (anweiss) [reply src/validation/cbor.rs]: Ah yep. Good catch. Can leave as is then.

**Comment by anweiss** [comment]
@ericseppanen I 100% agree on re-factoring the validation logic. It is super confusing. My first attempt was all about getting it to work. For now, I'll merge this in and open a separate issue to track re-factoring the validation code.

**Pull request merged**